### ISSUE TITLE
How to convert a .csv to .fasta?

### ISSUE THREAD

**Issue by restivve** [joey711/phyloseq#1111]
Do you know of a simple way to convert my microbiome taxa file (in .csv format) to a .fna file for use in Qiime2? 

I want to create a tree in Qiime2 (using a script that is similar to this: http://qiime.org/scripts/filter_tree.html) and compare it to the tree generated by phyloseq. 

Thank you,
Victoria

**Comment by mikemc** [comment]
Victoria, we will need more information about your .csv file to be able to answer your question. How are the sequences contained in the csv file? The best/easiest way to save a fasta file from inside R is often to create a DNAStringSet object and then use the writeXStringSet() function using the Biostrings package. To create the DNAStringSet object, you want a list of sequences named by the taxon name.

**Comment by restivve** [comment]
I'm trying to convert my taxa file (created using 16s rRNA DNA, from the v3-v4 region and assigned to ASVs using DADA2 and taxonomy assigned through SILVA 1.3.2), which is in a .csv file (see attached screen shot for format) to a .fna file. 

Are you saying to create a vector of the sequences? Right now they are just listed like: CCTTGGTTCC...

**Comment by restivve** [comment]
![Uploading Screen Shot 2019-04-09 at 3.13.22 PM.png…]()
Here's the attachment!

**Comment by restivve** [comment]
![Screen Shot 2019-04-09 at 3 13 22 PM](https://user-images.githubusercontent.com/44236616/55828505-945cf080-5ada-11e9-8f38-1a0d8f974c4e.png)

**Comment by mikemc** [comment]
Yes, you need to read the csv table into R into a data frame. See `help(read.csv)` or https://r4ds.had.co.nz/data-import.html. Then you need to get a vector of the sequences to feed to `DNAStringSet()`. Run `help(Biostrings::DNAStringSet)` to see the help for the `Biostrings::DNAStringSet()` function. Then, you can use `Biostrings::writeXStringSet()` to write the .fasta file.

**Comment by restivve** [comment]
I am working in phyloseq! However, my samples were processed by the sequencing facility in DADA2 and provided to me in a .csv file. Am I still able to use Biostrings and just take the first column sequences? 

That first column also doesn't have a name, so I'm not sure how to call the column.

Victoria

**Comment by restivve** [comment]
This is what I've done:

```{r}
setwd("~/Desktop/October_2018_data/")
path ="~/Desktop/October_2018_data/"
getwd()

taxfile = "taxa_KKfish_silva132.csv"
tax_df = read.csv(taxfile)

sequences <- tax_df$X
Kingdom <- tax_df$Kingdom
Phylum <- tax_df$Phylum
Class <- tax_df$Class
Order <- tax_df$Order
Family <- tax_df$Family
Genus <- tax_df$Genus

library("Biostrings")

DNA_sequences <- DNAStringSet(x=sequences, use.names=TRUE)
```

But I don't think this is what you want because they are now all separate vectors.

**Comment by mikemc** [comment]
Sorry, do you want just the sequences in a fasta file, or do you want the taxonomy information in there somehow as well? I don't think you said you wanted the taxonomy information in there. You might take a look [here](https://github.com/joey711/phyloseq/issues/1114) for an example of including the taxonomy information in the sequence headers of the fasta file once you have a phyloseq object. If what you want is some QIIME2 specific format of specifying taxonomy in the fasta file though, you are better off asking on the QIIME2 forum

**Comment by ptpell77** [comment]
Mike,
I have sequences stored as refseq, and stored headers to ASV integers as you suggest. Unfortunately your script as written, "Biostrings::writeXStringSet(path_to_save_to, refseq(ps))" is not working for me. 

When I call it, Refseq produces the followingt: DNAStringSet object of length 1452827. With a "width' "seq" and "name" column/slot. But I get the err:  'x' must be an XStringSet object

Thanks!

**Comment by mikemc** [comment]
@ptpell77 GlobalPatterns does not have reference sequences 
```
> library(phyloseq)
> data(GlobalPatterns)
> GlobalPatterns
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]
> refseq(GlobalPatterns)
Error in access(physeq, "refseq", errorIfNULL) : refseq slot is empty.
```
so this is not expected to work. `Biostrings::writeXStringSet` can only be applied to an actual "XStringSet" object (usually a "DNAStringSet" object, specifically, is what will be relevant for microbiome data). You need to make sure that `refseq(ps)` actually gives you an XStringSet object before trying to pass it to `Biostrings::writeXStringSet`
